File: sdks/ts/memory/src/ontology/tag.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Entity tagging for chunk metadata enrichment. Scans chunk content
 * for mentions of known ontology entity types and attaches type
 * metadata for retrieval enrichment.
 *
 * Port of go/ontology/tag.go.
 */

import type { ResolvedOntology } from './store.js'
import type { ClassificationResult } from './classify.js'
import { determineCategory } from './classify.js'

/**
 * ChunkTag is type metadata attached to a chunk for retrieval
 * enrichment.
 */
export type ChunkTag = {
  readonly entityTypes?: readonly string[]
  readonly businessCategory?: string
  readonly confidence: number
  readonly documentClass: string
}

/**
 * Creates a ChunkTag for a given chunk based on the classification
 * result and resolved ontology. Scans the chunk content for mentions
 * of known entity types from the ontology.
 */
export function tagChunk(
  content: string,
  classification: ClassificationResult,
  ontology: ResolvedOntology | undefined,
): ChunkTag {
  let businessCategory = classification.category
  let entityTypes: string[] | undefined

  if (ontology !== undefined) {
    const lower = content.toLowerCase()
    const matched = matchEntityTypes(lower, ontology)
    if (matched.length > 0) {
      entityTypes = matched
    }

    const ontologyCategory = determineCategory(content, ontology)
    if (ontologyCategory !== 'general') {
      businessCategory = ontologyCategory
    }
  }

  return {
    entityTypes,
    businessCategory,
    confidence: classification.confidence,
    documentClass: classification.class,
  }
}

/**
 * Creates ChunkTags for a batch of chunks. Each chunk is
 * independently tagged against the classification and ontology.
 */
export function tagChunks(
  contents: readonly string[],
  classification: ClassificationResult,
  ontology: ResolvedOntology | undefined,
): ChunkTag[] {
  return contents.map((content) => tagChunk(content, classification, ontology))
}

function matchEntityTypes(
  lowerContent: string,
  ontology: ResolvedOntology,
): string[] {
  const seen = new Set<string>()
  const matched: string[] = []

  for (const nt of ontology.nodeTypes) {
    const dotIdx = nt.type.indexOf('.')
    if (dotIdx < 0) continue
    const name = nt.type.slice(dotIdx + 1)
    const spaced = name.replace(/_/g, ' ')

    if (lowerContent.includes(name) || lowerContent.includes(spaced)) {
      if (!seen.has(nt.type)) {
        seen.add(nt.type)
        matched.push(nt.type)
      }
    }
  }

  return matched
}

File: go/ontology/tag.go
// SPDX-License-Identifier: Apache-2.0
package ontology

import "strings"

// ChunkTag is type metadata attached to a chunk for retrieval
// enrichment. It records which ontology entity types are present,
// the dominant business category, classification confidence, and
// the document class.
type ChunkTag struct {
	EntityTypes      []string `json:"entityTypes,omitempty"`
	BusinessCategory string   `json:"businessCategory,omitempty"`
	Confidence       float64  `json:"confidence"`
	DocumentClass    string   `json:"documentClass"`
}

// TagChunk creates a ChunkTag for a given chunk based on the
// classification result and resolved ontology. It scans the chunk
// content for mentions of known entity types from the ontology and
// propagates the document classification's business category.
func TagChunk(content string, classification ClassificationResult, ontology *ResolvedOntology) ChunkTag {
	tag := ChunkTag{
		BusinessCategory: classification.Category,
		Confidence:       classification.Confidence,
		DocumentClass:    string(classification.Class),
	}

	if ontology == nil {
		return tag
	}

	lower := strings.ToLower(content)
	entityTypes := matchEntityTypes(lower, ontology)
	if len(entityTypes) > 0 {
		tag.EntityTypes = entityTypes
	}

	// Override category if ontology voting produces a stronger signal
	ontologyCategory := DetermineCategory(content, ontology)
	if ontologyCategory != "general" {
		tag.BusinessCategory = ontologyCategory
	}

	return tag
}

// matchEntityTypes scans lowercased content for mentions of known
// ontology node types and returns a deduplicated list of matched type
// identifiers.
func matchEntityTypes(lowerContent string, ontology *ResolvedOntology) []string {
	seen := make(map[string]struct{})
	matched := make([]string, 0)

	for _, nt := range ontology.NodeTypes {
		parts := strings.SplitN(nt.Type, ".", 2)
		if len(parts) != 2 {
			continue
		}
		name := parts[1]

		// Match against both snake_case and space-separated forms
		spaced := strings.ReplaceAll(name, "_", " ")
		if strings.Contains(lowerContent, name) || strings.Contains(lowerContent, spaced) {
			if _, ok := seen[nt.Type]; !ok {
				seen[nt.Type] = struct{}{}
				matched = append(matched, nt.Type)
			}
		}
	}

	return matched
}

// TagChunks creates ChunkTags for a batch of chunks. Each chunk is
// independently tagged against the classification and ontology.
func TagChunks(contents []string, classification ClassificationResult, ontology *ResolvedOntology) []ChunkTag {
	tags := make([]ChunkTag, len(contents))
	for i, content := range contents {
		tags[i] = TagChunk(content, classification, ontology)
	}
	return tags
}
